docs/rfc/rfc-105-event-observer.md

```diff
@@ -0,0 +1,195 @@
+For consensus state, this may end up looking something like:
+
+```golang
+package consensus
+
+// StateObserver is specific to the consensus.State struct, and all of its
+// methods are called from within consensus.State instead of using an "event
+// bus". This allows for greater compile-time guarantees through doing away with
+// the generic pub/sub mechanism in the event bus.
+//
+// Note how all methods are infallible (i.e. they do not return any errors).
+// This is functionally equivalent to the fire-and-forget pattern implemented by
+// the event bus.
+//
+// Also note how method names are prefixed by the name of the relevant producer
+// of events (in this case "ConsensusState", corresponding to the
+// consensus.State struct). This is intentional to allow composition of
+// observers of multiple different components without function names clashing.
+//
+// Finally, given that this is just straightforward Go, it is up to either the
+// caller or the callee to decide how to handle the concurrency of certain
+// events. The event bus approach, by contrast, is always concurrent and relies
+// on Go channels, which could end up filling up and causing back-pressure into
+// the caller (already observed in slow WebSocket subscribers).
+type StateObserver interface {
+    ConsensusStateNewRoundStep(ev EventDataRoundState)
+    ConsensusStateTimeoutPropose(ev EventDataRoundState)
+    ConsensusStateTimeoutWait(ev EventDataRoundState)
+    // ...
+}
+```
+
+And on `consensus.State` one could easily either supply such an observer in the
+constructor, or define a new method that allows one to set the observer, e.g.:
+
+```golang
+package consensus
+
+func (cs *State) SetObserver(obs StateObserver) {
+    cs.observer = obs
+}
+```
+
+Then, instead of publishing events via the event bus, one would simply do the
+following:
+
+```diff
+-        if err := cs.eventBus.PublishEventNewRoundStep(rs); err != nil {
+-            cs.Logger.Error("failed publishing new round step", "err", err)
+-        }
++        // Notify the observer
++        cs.observer.ConsensusStateNewRoundStep(rs)
+```
+
+### Fanout Observers
+
+How then does one simulate the same sort of pub/sub functionality that the
+`EventBus` provides using this approach? Where this sort of behaviour is
+absolutely necessary, it is trivial to build a "fanout" observer that implements
+this interface, e.g.:
+
+```golang
+type StateFanoutObserver struct {
+    observers []StateObserver
+}
+
+func NewStateFanoutObserver(observers ...StateObserver) *StateFanoutObserver {
+    return &StateFanoutObserver{
+        observers: observers,
+    }
+}
+
+func (o *StateFanoutObserver) ConsensusStateNewRoundStep(ev EventDataRoundStep) {
+    for _, obs := range o.observers {
+        obs.ConsensusStateNewRoundStep(ev)
+    }
+}
+
+// ...
+```
+
+### Testing
+
+Many tests in the CometBFT codebase rely heavily on subscribing to specific
```

Comment:
The RFC covers very well the _publisher_ side of coupled components, but not so much the _subscriber_ side, apart from the discussion here about testing. 

Looking through the codebase, I see these two types of subscriptions possible:
- [Subscribe](https://github.com/cometbft/cometbft/blob/v0.38.x/types/event_bus.go#L75)
- [SubscribeUnbuffered](https://github.com/cometbft/cometbft/blob/v0.38.x/types/event_bus.go#L86)

So I have two remarks/suggestions:

1. It might help to clarify if both possible of subscriptions are possible with event observers (I understood they are).
2. [Nit] It seems like the heaviest users of subscription are test modules: more than half of references I found to the above subscribe methods originate from _test.go files, which was surprising to me. It seems useful to clarify that part of the motivation of this change is to simplify testing (which we do at the end of the RFC but not as motivation at the outset).

Reply:
I've updated the RFC. The two subscribe mechanisms are pretty much the same - the only difference is the channel capacity.